product_multi_company/migrations/13.0.1.1.0/post-migration.py

```diff
@@ -0,0 +1,24 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+import logging
+
+from openerp import SUPERUSER_ID, api
+
+_logger = logging.getLogger(__name__)
+
+
+def migrate(cr, installed_version):
+    _logger.debug(
+        "Populate new multicompany field on product.product which we added "
+        "because it was not correctly inherited from product.template."
+    )
+    env = api.Environment(cr, SUPERUSER_ID, {})
+    for product_product in env["product.product"].search([]):
+        if len(product_product.company_ids) == 0:
```

Comment:
This is not going to work. You need to read template model.

Reply:
In the 13.0.1.0.0 version `product_product.company_ids` is a related field and would not always == 0.
However, the upgrade runs first. This makes `product_product.company_ids` a real field and adds the `product_product_res_company_rel` table. The default is not set even if the system user has companies (observed by testing). This may be a side effect of what you mention below (the table does not exist), but in this case I would expect an error, and I don't see that during testing (not even a warning).
After the `product_product_res_company_rel` table is added the post-migration script runs and since no writes were performed the table is still empty. It is subsequently populated by the write in the post-migration script.

I tested this by creating a new database in a clean docker environment with the 13.0.1.0.0 version of this module. I added companies, users, products and product variants and added companies on them. I then reproduced the original bug (#252).
I then checked out the 13.0.2.0.0 version and commented out all other ways `product_product.company_ids` could be written to in that module. I also added a set_trace and observed that `product_product.company_ids` was indeed written in the migration script.

That said (long breath), I'm eager to learn if I overlooked something. These _inherits and abstract models can set me on the wrong trail some times. :sweat_smile: 

I did find a mistake in product_template.py. I'll push a new commit shortly.